### PR TITLE
Type Checking in CI

### LINKED ISSUE #3699: make type checking in CI blocking
This issue depends on #3698 which requests mypy and pyright to be run in CI to check type annotations. This issue suggests making the job non-blocking as type annotation is not yet implemented. Once the annotations are implemented to a point where the checking job passes, it should be made blocking to avoid breaking the existing annotations and to avoid introducing new broken ones.

This is part of @umak1106's outreach project.

### PR DESCRIPTION
**Pull request by umak1106**
Fixes #3699 #3698 

Changes made in this Pull Request:
 - Added mypy.ini and mypy_requirements.txt

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Comment by umak1106** [comment]
Hello mentors ,
I have added the mypy.ini file . Currently I have set it to ignore all the errors in the modules . As and when I'll start type hinting the module I'll make the required changes to the file .
Incase you have any suggestions regarding the improvements to be made here kindly let me know

File: .github/workflows/mypy.yaml
Comment (jbarnoud) [inline .github/workflows/mypy.yaml]: We do not have a master branch anymore for MDAnalysis. You can read the discussion about this here: https://github.com/MDAnalysis/mdanalysis/issues/3542

File: .github/workflows/mypy.yaml
Comment (jbarnoud) [inline .github/workflows/mypy.yaml:21]: Why do you need this conditional? Note, I am not saying you do not need it, I genuinely do not know about the possible usefulness of that line.

File: .github/workflows/mypy.yaml
Comment (jbarnoud) [inline .github/workflows/mypy.yaml]: Out of curiosity, why did you choose this version of python?

**Review by jbarnoud — COMMENTED** [review]
It looks like there is a syntax error in `mypy.yaml`. You can see the result of the workflow here: https://github.com/umak1106/mdanalysis/actions

File: .github/workflows/mypy.yaml
Comment (lilyminium) [reply .github/workflows/mypy.yaml:21]: This limits this action only to pushes and PRs to `MDAnalysis/mdanalysis`, instead of forks. We have it throughout the rest of our CI, I assume as a courtesy to the GitHub servers / personal allotments.

Comment (IAlibay) [reply .github/workflows/mypy.yaml:21]: it's a nice way to avoid forks from running CI if they don't want to (for mypy it's not that bad, but every little helps when it comes to reducing CO2 emissions) edit: Lily got there before me :P

File: .github/workflows/mypy.yaml
Comment (IAlibay) [inline .github/workflows/mypy.yaml]: Please use the grouping which we use in gh-ci.yaml This one is prone to collisions

File: .github/workflows/mypy.yaml
Comment (jbarnoud) [inline .github/workflows/mypy.yaml]: In the main workflow, we use version 3.

**Review by jbarnoud — COMMENTED** [review]
Good progress today! The workflow runs :smiley: 

I see a couple of things left to fix but we have a great start here. One thing to fix is that you need the workflow to install the dependencies fo MDAnalysis. In the main workflow, it is done here: https://github.com/MDAnalysis/mdanalysis/blob/ec5da1de682322c436970252f03b5a24ea673113/.github/workflows/gh-ci.yaml#L91

File: .github/workflows/mypy.yaml
Comment (umak1106) [reply .github/workflows/mypy.yaml]: I removed master branch

File: .github/workflows/mypy.yaml
Comment (umak1106) [reply .github/workflows/mypy.yaml]: Actually in gh-ci both pylint and pypi used python version 3.8 so i did the same here

**Comment by umak1106** [comment]
@jbarnoud 
Hello mentor , I made some changes in the mypy.yaml file and am unable to figure out why some of the tests are failing . If i need to make any other changes please let me know.

**Comment by jbarnoud** [comment]
The tests being broken seems unrelated. We will look at fixing this.

Regarding the install phase in the yaml file, I do not think we need the full matrix. Indeed, the type check should not change with different operating systems, for instance. This means you should be able to simplify your file by removing the matrix you copied and, instead, use values directly where you have variables. I *think* you should go with `true` for the full dependencies, `1.19.0` for the NumPy version (the oldest version we support), and I do not think you need extra pip dependencies.

As a side note, it would be good to have more informative commit messages. For the moment, all your commit messages are the same, which makes it difficult for me to know what you changed and why you changed it, and for you to look back at a specific commit. Note that we, as a project, are not very picky with commit messages; other projects, though, may use commit messages almost like cover letters where a contributor justifies why the commit should be included. As an example, your previous commit could have been titled something like "Use a dependency matrix for the mypy workflow".

File: .github/workflows/mypy.yaml
Comment (jbarnoud) [inline .github/workflows/mypy.yaml:36]: Looks like you need `mamba-version: "*"` for mamba to be installed. See https://github.com/MDAnalysis/mdanalysis/blob/ec5da1de682322c436970252f03b5a24ea673113/.github/workflows/gh-ci.yaml#L88

File: .github/workflows/mypy.yaml
Comment (IAlibay) [inline .github/workflows/mypy.yaml:42]: Given that the 2.3.0 release will likely support a minimum numpy version of 1.20 and that type annotations were added to numpy in 1.20, it would be better to set this to 1.20.0.

I'll submit a PR today or tomorrow raising our minimum numpy requirements to 1.20 for everything else.

File: .github/workflows/mypy.yaml
Comment (umak1106) [reply .github/workflows/mypy.yaml:36]: yes  mentor , I just realized after making the commit .

File: .github/workflows/mypy.yaml
Comment (umak1106) [reply .github/workflows/mypy.yaml:42]: ok mentor.

File: .github/workflows/mypy.yaml
Comment (jbarnoud) [reply .github/workflows/mypy.yaml:36]: I'll wait a bit before giving away too many answers, then :wink:

Comment (umak1106) [reply .github/workflows/mypy.yaml:36]: Actually am getting the error , nothing provides numpy 1.20.0 . What am I supposed to do about it ?

Comment (jbarnoud) [reply .github/workflows/mypy.yaml:36]: The problem is the way you specify the numpy version. It is a bit unintuitive: the `numpy` attribute expects `numpy=1.20.0` and not just `1.20.0`.

Since, we are at it, you should also check the `channel` attribute.

Comment (jbarnoud) [reply .github/workflows/mypy.yaml:36]: Oh, you already checked `channel` :smile:

Comment (umak1106) [reply .github/workflows/mypy.yaml:36]: I think the workflow runs properly now 😄.

**Review by jbarnoud — APPROVED** [review]
Looks like it is working :tada: 

I cannot merge right now, until the rest of the CI is fixed. But we will fix that soon.

Also, @IAlibay could you have a quick look? I am not super up to date with our CI so I may be missing things.

**Review by richardjgowers — APPROVED** [review]
LGTM.  Since the CI entry from this isn't failing, I don't think merging this makes the current CI problems any worse...

File: .github/workflows/mypy.yaml
Comment (IAlibay) [inline .github/workflows/mypy.yaml:47]: Do you not need to install MDAnalysis?

**Review by IAlibay — COMMENTED** [review]
Sorry only doing a very cursory look at things, overall looks good but I don't remember if you have to (or should) install the MDA package for a mypy check. If so then a `python setup.py install` call somewhere (unless I missed it), would be necessary

**Comment by IAlibay** [comment]
Also I agree with @richardjgowers it's fine to go ahead with the merge, the failures are unrelated.

File: .github/workflows/mypy.yaml
Comment (umak1106) [reply .github/workflows/mypy.yaml:47]: Umm as we already installed the dependencies should i add 
          cd package
          python setup.py develop --no-deps
instead of 
cd package
python setup.py install ?

File: .github/workflows/mypy.yaml
Comment (tylerjereddy) [inline .github/workflows/mypy.yaml:43]: For future reference, I'd suggest using even newer NumPy for the type checking because the type hint machinery available in NumPy continues to improve with each release. There are some "new in version.." statements in the docs here:

https://numpy.org/devdocs/reference/typing.html#typing-numpy-typing

File: mypy.ini
Comment (tylerjereddy) [inline mypy.ini:3]: If the team decides that using NumPy `1.22.x` series is acceptable for the purposes of type checking, you can use the new NumPy plugin as well: `plugins = numpy.typing.mypy_plugin`

see: https://numpy.org/devdocs/reference/typing.html#mypy-plugin

**Review by tylerjereddy — APPROVED** [review]
Sounds like folks are happy here, I'll squash merge then given some of the commit messages.

I added a few comments about fancy things you could use if you bump the NumPy version a bit more, but I'll leave that up to the team/mentors.

**Comment by tylerjereddy** [comment]
I created a "typing" label and applied it here, hope that's ok, feel free to change it if you prefer something else.

**Pull request merged**
